apm-sniffer/apm-toolkit-activation/apm-toolkit-common/src/main/java/org/apache/skywalking/apm/toolkit/common/log/GRPCLogReportServiceClient.java

```diff
@@ -0,0 +1,140 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package org.apache.skywalking.apm.toolkit.common.log;
+
+import java.util.List;
+import java.util.Objects;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+import org.apache.skywalking.apm.agent.core.boot.OverrideImplementor;
+import org.apache.skywalking.apm.agent.core.conf.Config;
+import org.apache.skywalking.apm.agent.core.logging.api.ILog;
+import org.apache.skywalking.apm.agent.core.logging.api.LogManager;
+import org.apache.skywalking.apm.agent.core.remote.GRPCStreamServiceStatus;
+import org.apache.skywalking.apm.agent.core.remote.LogReportServiceClient;
+import org.apache.skywalking.apm.agent.core.util.CollectionUtil;
+import org.apache.skywalking.apm.commons.datacarrier.DataCarrier;
+import org.apache.skywalking.apm.commons.datacarrier.buffer.BufferStrategy;
+import org.apache.skywalking.apm.network.common.v3.Commands;
+import org.apache.skywalking.apm.network.logging.v3.LogData;
+import org.apache.skywalking.apm.network.logging.v3.LogReportServiceGrpc;
+
+import io.grpc.ManagedChannel;
+import io.grpc.ManagedChannelBuilder;
+import io.grpc.StatusRuntimeException;
+import io.grpc.stub.StreamObserver;
+
+/**
+ * Report log to server by grpc
+ */
+@OverrideImplementor(LogReportServiceClient.class)
+public class GRPCLogReportServiceClient extends LogReportServiceClient {
+
+    private static final ILog LOGGER = LogManager.getLogger(GRPCLogReportServiceClient.class);
+
+    private volatile DataCarrier<LogData> carrier;
+
+    private LogReportServiceGrpc.LogReportServiceStub asyncStub;
+
+    private ManagedChannel channel;
+
+    private AtomicBoolean disconnected = new AtomicBoolean(false);
+
+    @Override
+    public void boot() throws Throwable {
+        carrier = new DataCarrier<>("gRPC-log", "gRPC-log", Config.Buffer.CHANNEL_SIZE, Config.Buffer.BUFFER_SIZE,
+                BufferStrategy.IF_POSSIBLE);
+        carrier.consume(this, 1);
+        channel = ManagedChannelBuilder
+                .forAddress(GRPCLogConfig.Plugin.GRPCLog.SERVER_HOST, GRPCLogConfig.Plugin.GRPCLog.SERVER_PORT)
+                .usePlaintext().build();
+        asyncStub = LogReportServiceGrpc.newStub(channel)
+                .withMaxOutboundMessageSize(GRPCLogConfig.Plugin.GRPCLog.MAX_MESSAGE_SIZE);
+    }
+
+    @Override
+    public void shutdown() {
+        try {
+            if (channel != null) {
+                channel.shutdownNow();
+            }
+            carrier.shutdownConsumers();
+        } catch (Throwable t) {
+            LOGGER.error(t.getMessage(), t);
+        }
+    }
+
+    @Override
+    public void produce(LogData logData) {
+        if (Objects.nonNull(logData) && !carrier.produce(logData)) {
+                LOGGER.warn("LogReportServiceClient discard logData: " + logData.toString());
```

Comment:
Format the indents

Reply:
I think print all discarded logs would be very dangerous, which could impact the system performance a lot. Please take a look how trace sender did. Don't print the log content.